### PR TITLE
PTR-27 Check circular dependency. Add java docs for some interfaces

### PR DESCRIPTION
**Pull request by mykolad4**

### CODE REVIEW

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java [hunk 1]
```diff
@@ -102,19 +104,9 @@ private <T> T createBean(BeanDefinition beanDefinition) {
...
-                                        .map(this::create)
-                                        .orElseThrow(
-                                                () -> new BeanCreationException("Could not create bean with type: " + dependsOnClass.getName())
-                                        );
-                    } else {
-                        beanToInject = this.getOptionalBean(dependsOnClass).get();
-                    }
+                    Object beanToInject = this.getOptionalBean(dependsOnClass).isPresent()
+                            ? this.getOptionalBean(dependsOnClass).get()
+                            : createDependentBean(dependsOnClass);
```
Comment (mitssev) [inline src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:109]: probably better avoid isPresent and jist use orElseGet

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java [hunk 2]
```diff
@@ -128,6 +120,27 @@ private <T> T createBean(BeanDefinition beanDefinition) {
...
+                .orElseThrow(
+                        () -> new BeanCreationException("Could not create bean with type: " + dependsOnClass.getName())
+                );
+        circularDependencies.remove(dependsOnClass.getSimpleName());
+        return dependsOnClass.cast(o);
+    }
+
+    private void checkCircularDependency(String name) {
+        if (circularDependencies.contains(name)) {
+            circularDependencies.add(name + "!");
```
Comment (mitssev) [inline src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:137]: didnt get this, adding duplicated name but with exclamation?

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java (hunk 1 above)
Comment (mykolad4) [reply src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:109]: I didn't manage to do it:(

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java (hunk 2 above)
Comment (mykolad4) [reply src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:137]: It's been done in order to make exception message more understandable. There will be something like "Could not initialize application. Circular dependencies are found:  A -> B -> C -> D -> A!" in console logs

**Review by mitssev — APPROVED** [review]

**Review by petrosteam — APPROVED** [review]

**Pull request merged**